server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java

```diff
@@ -610,7 +651,212 @@ public void run() {
+            String dir = mountUri(uri, nfsVersion);
+            return _parent + "/" + dir;
+        } catch (Exception e) {
+            String msg = "GetRootDir for " + secUrl + " failed due to " + e.toString();
+            LOGGER.error(msg, e);
+            throw new CloudRuntimeException(msg);
+        }
+    }
+
+    protected String mountUri(URI uri, Integer nfsVersion) throws UnknownHostException {
+        String uriHostIp = getUriHostIp(uri);
+        String nfsPath = uriHostIp + ":" + uri.getPath();
+
+        String dir = UUID.nameUUIDFromBytes(nfsPath.getBytes(com.cloud.utils.StringUtils.getPreferredCharset())).toString();
+        String localRootPath = _parent + "/" + dir;
+
+        String remoteDevice;
+        if (uri.getScheme().equals("cifs")) {
+            remoteDevice = "//" + uriHostIp + uri.getPath();
+            LOGGER.debug("Mounting device with cifs-style path of " + remoteDevice);
+        } else {
+            remoteDevice = nfsPath;
+            LOGGER.debug("Mounting device with nfs-style path of " + remoteDevice);
+        }
+        mount(localRootPath, remoteDevice, uri, nfsVersion);
+        return dir;
+    }
+
+    protected String getUriHostIp(URI uri) throws UnknownHostException {
+        String nfsHost = uri.getHost();
+        InetAddress nfsHostAddr = InetAddress.getByName(nfsHost);
+        String nfsHostIp = nfsHostAddr.getHostAddress();
+        LOGGER.info("Determined host " + nfsHost + " corresponds to IP " + nfsHostIp);
+        return nfsHostIp;
+    }
+
+    protected void mount(String localRootPath, String remoteDevice, URI uri, Integer nfsVersion) {
+        LOGGER.debug("mount " + uri.toString() + " on " + localRootPath + ((nfsVersion != null) ? " nfsVersion=" + nfsVersion : ""));
+        ensureLocalRootPathExists(localRootPath, uri);
+
+        if (mountExists(localRootPath, uri)) {
+            return;
+        }
+
+        attemptMount(localRootPath, remoteDevice, uri, nfsVersion);
+
+        checkForVolumesDir(localRootPath);
+    }
+
+    protected boolean checkForVolumesDir(String mountPoint) {
+        String volumesDirLocation = mountPoint + "/" + "volumes";
+        return createDir("volumes", volumesDirLocation, mountPoint);
+    }
+
+    protected boolean createDir(String dirName, String dirLocation, String mountPoint) {
+        boolean dirExists = false;
+
+        File dir = new File(dirLocation);
+        if (dir.exists()) {
+            if (dir.isDirectory()) {
+                LOGGER.debug(dirName + " already exists on secondary storage, and is mounted at " + mountPoint);
+                dirExists = true;
+            } else {
+                if (dir.delete() && _storage.mkdir(dirLocation)) {
+                    dirExists = true;
+                }
+            }
+        } else if (_storage.mkdir(dirLocation)) {
+            dirExists = true;
+        }
+
+        if (dirExists) {
+            LOGGER.info(dirName + " directory created/exists on Secondary Storage.");
```

Comment:
you have obviously given log levels thoughts. let's discuss them. As it seems these two might not be of the same level.